fix: toggleList

ghingis opened the pull request:
List toggle had some issues when it came to changing type or revering it back to standard paragraphs this PR intends to fix those issues and it has a couple of test for the different usecases.

* Normalizer now merges lists with the same type next to each other also if a list has no lic and it has children it moves those childrens up a level 
* toggleList works as expected (check tests)
* ToolbarElement now uses the provided active prop
* ToolbarList leverages this since the default ToolbarElement solution would activate both button if there is an unordered list within a ordered list.
* moveListItemDown wrap transformations in withoutNormalizing (it caused a pathing issue since the normalization would remove the created empty list)

dylans commented on packages/elements/list/src/transforms/toggleList.ts:
```diff
@@ -1,46 +1,139 @@
+      }
+    } else {
+      // selection is a range
+      const [statPoint, endPoint] = Range.edges(editor.selection!);
```
```suggestion
      const [startPoint, endPoint] = Range.edges(editor.selection!);
```

dylans commented on packages/elements/list/src/transforms/toggleList.ts:
```diff
@@ -1,46 +1,139 @@
+
+      const commonEntry: NodeEntry<Node> = Node.common(
+        editor,
+        statPoint.path,
```
```suggestion
        startPoint.path,
```

dylans reviewed (changes requested):
large PR so I haven't had time to review it, but here's an obvious typo.

dylans commented on packages/elements/list/src/transforms/unwrapList.ts:
```diff
@@ -1,30 +1,35 @@
-    at,
-    match: { type: getPlatePluginType(editor, ELEMENT_LI) },
-  });
+    unwrapNodes(editor, {
```
should the unwrap calls get wrapped in withoutNormalizing calls?

dylans commented on packages/elements/list/src/transforms/toggleList.spec.tsx:
```diff
@@ -0,0 +1,348 @@
+});
+
+describe('toggle off', () => {
+  it('should spit a simple list to two', () => {
```
```suggestion
  it('should split a simple list to two', () => {
```

dylans commented on packages/elements/list/src/transforms/toggleList.spec.tsx:
```diff
@@ -0,0 +1,348 @@
+    expect(input.children).toEqual(output.children);
+  });
+
+  it('should only toggle everything that selected', () => {
```
```suggestion
  it('should only toggle everything that is selected', () => {
```

dylans reviewed (commented):
One more question, are there are test scenarios that we could add for the normalization step that would have failed with the old code that are now passing?

nemanja-tosic commented on packages/elements/list/src/transforms/toggleList.ts:
```diff
@@ -1,46 +1,139 @@
 
 export const toggleList = (editor: SPEditor, { type }: { type: string }) => {
-  if (!editor.selection) return;
+  if (!editor.selection) {
```
Moving this check inside of Editor.withoutNormalizing has 2 benefits:

1. all as Range casts can be removed
2. brackets can be removed from toggleList, decreasing indentation

nemanja-tosic commented on packages/elements/list/src/transforms/toggleList.ts:
```diff
@@ -1,46 +1,139 @@
+      // selection is a range
+      const [statPoint, endPoint] = Range.edges(editor.selection!);
+
+      const commonEntry: NodeEntry<Node> = Node.common(
```
Node is the default.

```suggestion
      const commonEntry: NodeEntry = Node.common(
```

nemanja-tosic commented on packages/elements/list/src/normalizers/normalizeListItem.ts:
```diff
@@ -90,6 +94,26 @@ export const normalizeListItem = (
+      const parent = getParent(editor, listItem[1]);
+      const sublist = firstLiChild;
+      const children = getChildren(firstLiChild).reverse();
+      Editor.withoutNormalizing(editor, () => {
```
withoutNormalizing, on the other hand, is not needed in normalizations (the other call should go too).

ghingis replied:
you are right `normalizeNode` is always called from a `withoutNormalizing`, I'll remove them

zbeyens reviewed (commented):
Excellent PR, left a few comments

zbeyens commented:
As @dylans mentioned, testing sibling lists would be interesting

ghingis commented on packages/elements/list/src/normalizers/getListNormalizer.ts:
```diff
@@ -40,9 +44,33 @@ export const getListNormalizer = (
 
-      if (normalizeNestedList(editor, { nestedListItem: [node, path] })) {
-        return;
+      const nextPath = [...path.slice(0, -1), path[path.length - 1] + 1];
```
It seems it does the same, I'll change it..

ghingis commented on packages/elements/list/src/normalizers/getListNormalizer.ts:
```diff
@@ -40,9 +44,33 @@ export const getListNormalizer = (
+        });
+      }
+
+      if (path[path.length - 1] > 0) {
```
yeah, with the getNode error tolerance and with a casting to Path it can be removed.. it just feels a bit wrong :)

ghingis commented on packages/elements/list/src/normalizers/getListNormalizer.ts:
```diff
@@ -40,9 +44,33 @@ export const getListNormalizer = (
+        }
       }
+
+      normalizeNestedList(editor, { nestedListItem: [node, path] });
```
me neither, I probably had my reasons but checking the code they were probably wrong, I put it back..

ghingis commented:
I found that some `withoutNormalizing` was still missing, fixed the typos and changed things as you recommended.
I also added some test for the normalizers.
